File: 404.SumofLeftLeaves.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumOfLeftLeaves = function(root) {
    if (root) {
        if (root.left && !root.left.left && !root.left.right) {
            return root.left.val + sumOfLeftLeaves(root.right);
        }
        return sumOfLeftLeaves(root.left) + sumOfLeftLeaves(root.right);
    }
    return 0;
};

var sumOfLeftLeaves = function(root) {
    let sum = 0;
    const stack = [root];
    while (stack.length) {
        const node = stack.pop();
        if (node) {
            if (node.left && !node.left.left && !node.left.right) {
                sum += node.left.val;   
            } else {
                stack.push(node.left);   
            }
            stack.push(node.right);
        }
    }
    return sum;
};

File: Golang/Tree/404.Sum-of-left-leaves.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func sumOfLeftLeavesRecursive(root *TreeNode) int {
	ret := 0
	var recursive func(node *TreeNode)
	recursive = func(node *TreeNode) {
		if node != nil {
			if node.Left != nil && node.Left.Left == nil && node.Left.Right == nil {
				ret += node.Left.Val
			}
			recursive(node.Left)
			recursive(node.Right)
		}
	}
	recursive(root)
	return ret
}

func sumOfLeftLeavesStack(root *TreeNode) int {
	ret := 0
	if root != nil {
		stack := []*TreeNode{}
		node := root
		for {
			if node != nil {
				stack = append(stack, node)
				node = node.Left
			} else if len(stack) > 0 {
				node = stack[len(stack)-1]
				stack = stack[:len(stack)-1]
				if node.Left != nil && node.Left.Left == nil && node.Left.Right == nil {
					ret += node.Left.Val
				}
				node = node.Right
			} else {
				break
			}
		}
	}
	return ret
}
